src/class/audio/audio_device.h

```diff
@@ -0,0 +1,399 @@
+#endif
+
+// Size of control buffer used to receive and send control messages via EP0 - has to be big enough to hold your biggest request structure e.g. range requests with multiple intervals defined or cluster descriptors
+#ifndef CFG_TUD_AUDIO_CTRL_BUF_SIZE
+#error You must define an audio class control request buffer size!
+#endif
+
+// Use of TX/RX FIFOs - If sizes are not zero, audio.c implements FIFOs for RX and TX (whatever defined).
+// For RX: the input stream gets decoded into its corresponding channels, where for each channel a FIFO is setup to hold its data -> see: audio_rx_done_cb().
+// For TX: the output stream is composed from CFG_TUD_AUDIO_N_CHANNELS_TX channels, where for each channel a FIFO is defined.
+// Further, it implements encoding and decoding of the individual channels (parameterized by the defines below).
+// If you don't use the FIFOs you need to handle encoding and decoding on your own in audio_rx_done_cb() and audio_tx_done_cb(). This, however, allows for optimizations.
+
+#ifndef CFG_TUD_AUDIO_TX_FIFO_SIZE
+#define CFG_TUD_AUDIO_TX_FIFO_SIZE  0                           // Buffer size per channel
+#endif
+
+#if CFG_TUD_AUDIO_TX_FIFO_SIZE && CFG_TUD_AUDIO_TX_DMA_RINGBUFFER_SIZE
+#error TX_FIFOs and TX_DMA_RINGBUFFER can not be used simultaneously!
+#endif
+
+#ifndef CFG_TUD_AUDIO_RX_FIFO_SIZE
+#define CFG_TUD_AUDIO_RX_FIFO_SIZE  0                           // Buffer size per channel
+#endif
+
+// End point sizes - Limits: Full Speed <= 1023, High Speed <= 1024
+#ifndef CFG_TUD_AUDIO_EPSIZE_IN
+#define CFG_TUD_AUDIO_EPSIZE_IN 0   // TX
+#endif
+
+#ifndef CFG_TUD_AUDIO_EPSIZE_OUT
+#define CFG_TUD_AUDIO_EPSIZE_OUT 0  // RX
+#endif
+
+#ifndef CFG_TUD_AUDIO_ENABLE_FEEDBACK_EP
+#define CFG_TUD_AUDIO_ENABLE_FEEDBACK_EP 0                      // Feedback
+#endif
+
+#ifndef CFG_TUD_AUDIO_INT_CTR_EPSIZE_IN
+#define CFG_TUD_AUDIO_INT_CTR_EPSIZE_IN 0                       // Audio interrupt control
+#endif
+
+#if CFG_TUD_AUDIO_INT_CTR_EPSIZE_IN
+#ifndef CFG_TUD_AUDIO_INT_CTR_BUFSIZE
+#define CFG_TUD_AUDIO_INT_CTR_BUFSIZE   6                       // Buffer size of audio control interrupt EP - 6 Bytes according to UAC 2 specification (p. 74)
+#endif
+#endif
+
+#ifndef CFG_TUD_AUDIO_N_CHANNELS_TX
+#define CFG_TUD_AUDIO_N_CHANNELS_TX                     1
+#endif
+
+#ifndef CFG_TUD_AUDIO_N_CHANNELS_RX
+#define CFG_TUD_AUDIO_N_CHANNELS_RX                     1
+#endif
+
+// Audio data format types
+#ifndef CFG_TUD_AUDIO_FORMAT_TYPE_TX
+#define CFG_TUD_AUDIO_FORMAT_TYPE_TX                AUDIO_FORMAT_TYPE_UNDEFINED     // If this option is used, an encoding function has to be implemented in audio_device.c
+#endif
+
+#ifndef CFG_TUD_AUDIO_FORMAT_TYPE_RX
+#define CFG_TUD_AUDIO_FORMAT_TYPE_RX                AUDIO_FORMAT_TYPE_UNDEFINED     // If this option is used, a decoding function has to be implemented in audio_device.c
+#endif
+
+// Audio data format type I specifications
+#if CFG_TUD_AUDIO_FORMAT_TYPE_TX == AUDIO_FORMAT_TYPE_I
```

Comment:
Very well spotted! Do you think it is worth reworking the defines in audio_device.h such that `AUDIO_FORMAT_TYPE_I` is an enumeration again?

Reply:
Yeah, we could. But it can wait until we finalize the api. There is little different in the usage of macro and enum in this case. There is no hurry at all

Reply:
Ok!